File: agent-v01/scripts/generate-skill-profiles.py
#!/usr/bin/env python3
# ═══════════════════════════════════════════════════════════════
# generate-skill-profiles.py — Generate per-agent skill profiles
# (Tier 3) from SKILL-ROUTER.yaml (Tier 2).
#
# Each agent gets a compact profile at agent-v01/skills/profiles/
# listing ONLY its applicable skills — so an agent loads a small
# file instead of scanning inline tables.
#
# Usage: python3 agent-v01/scripts/generate-skill-profiles.py
# ═══════════════════════════════════════════════════════════════

import os
import sys

try:
    import yaml
except ImportError:
    print("ERROR: PyYAML required — pip install pyyaml")
    sys.exit(1)

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
ROUTER = os.path.join(ROOT, "agent-v01", "SKILL-ROUTER.yaml")
PROFILES_DIR = os.path.join(ROOT, "agent-v01", "skills", "profiles")

def main():
    if not os.path.exists(ROUTER):
        print(f"ERROR: {ROUTER} not found")
        sys.exit(1)

    with open(ROUTER) as f:
        data = yaml.safe_load(f)

    os.makedirs(PROFILES_DIR, exist_ok=True)
    router = data.get("router", {})

    for persona, phases in router.items():
        # Collect all skills for this persona
        core_skills = []
        additional = []
        conditional = {}

        for phase, sections in phases.items():
            core_skills.extend(sections.get("core", []))
            additional.extend(sections.get("additional", []))
            if "conditional" in sections:
                conditional.update(sections.get("conditional", {}))
            if "stacks" in sections:
                conditional.update({"stacks": list(sections["stacks"].keys())})
            if "security" in sections:
                additional.extend(sections.get("security", []))

        profile = {
            "persona": persona,
            "phases": list(phases.keys()),
            "core_skills": sorted(set(core_skills)),
            "additional_skills": sorted(set(additional)),
            "conditional_skills": conditional,
            "total_skills": len(set(core_skills + additional)),
            "load_strategy": "lazy — load SKILL.md only when task matches",
        }

        out_path = os.path.join(PROFILES_DIR, f"{persona}.yaml")
        with open(out_path, "w") as f:
            yaml.safe_dump(profile, f, sort_keys=False, default_flow_style=False)
        print(f"  ✅ {persona}: {profile['total_skills']} skills → {os.path.relpath(out_path, ROOT)}")

    print(f"\nProfiles generated in {os.path.relpath(PROFILES_DIR, ROOT)}")

if __name__ == "__main__":
    main()

File: agent-v01/scripts/generate-skill-profiles.rb
#!/usr/bin/env ruby
# ═══════════════════════════════════════════════════════════════
# generate-skill-profiles.rb — Generate per-agent skill profiles
# (Tier 3) from SKILL-ROUTER.yaml (Tier 2).
# Usage: ruby agent-v01/scripts/generate-skill-profiles.rb
# ═══════════════════════════════════════════════════════════════
require 'yaml'

ROOT = File.expand_path('../..', __dir__)
ROUTER = File.join(ROOT, 'agent-v01', 'SKILL-ROUTER.yaml')
PROFILES_DIR = File.join(ROOT, 'agent-v01', 'skills', 'profiles')

abort "ERROR: #{ROUTER} not found" unless File.exist?(ROUTER)

data = YAML.load_file(ROUTER)
require 'fileutils'; FileUtils.mkdir_p(PROFILES_DIR) unless Dir.exist?(PROFILES_DIR)

data['router'].each do |persona, phases|
  core = []
  additional = []
  conditional = {}

  phases.each do |phase, sections|
    core.concat(Array(sections['core']))
    additional.concat(Array(sections['additional']))
    conditional.merge!(sections['conditional']) if sections['conditional']
    conditional['stacks'] = sections['stacks'].keys if sections['stacks']
    additional.concat(Array(sections['security'])) if sections['security']
  end

  core = core.uniq.sort
  additional = additional.uniq.sort
  total = (core + additional).uniq.size

  profile = {
    'persona' => persona,
    'phases' => phases.keys,
    'core_skills' => core,
    'additional_skills' => additional,
    'conditional_skills' => conditional,
    'total_skills' => total,
    'load_strategy' => 'lazy — load SKILL.md only when task matches',
  }

  out = File.join(PROFILES_DIR, "#{persona}.yaml")
  File.write(out, profile.to_yaml)
  puts "  OK #{persona}: #{total} skills -> #{out}"
end

puts "\nProfiles generated in agent-v01/skills/profiles/"
